### PR TITLE
Automate creation of stubs

### LINKED ISSUE #169: Automate creation and update of stubs
With #162, we'll be using `stubConsoleMessages` in unit/integration tests. Currently, these are assembled by hand. However, we could write a script to automatically generate them. I think building off of https://github.com/janodvarko/prototypes will make it easy to write an npm script that builds the files for us.

### PR DESCRIPTION
**Pull request by linclark**
Fixes #169 
## Testing instructions
1. delete `stubs.js`
2. run `./mach test browser_webconsole_update_stubs.js`
3. run `npm test`

### CODE REVIEW

**Comment by linclark** [comment]
@bgrins I got it working with your suggestion for console api calls. However, I can't figure out what do to for `evaluationResult` or `pageError`. Do you have any thoughts?

**Comment by linclark** [comment]
This is a hot mess... but it works for evaluation results and page errors now. Will clean it up on Monday

**Comment by linclark** [comment]
@bgrins this code is gnarly, but I think its usefulness outweighs that.

Since it does its job, I suggest that we commit as is. We can create a follow-up issue to clean it up. We may end up changing the approach later on, so I don't think it's worth doing a thorough code review process for now. Let me know what you think, though.

File: devtools/client/webconsole/new-console-output/test/fixtures/stubs/pageError.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/* Any copyright is dedicated to the Public Domain.
+  http://creativecommons.org/publicdomain/zero/1.0/ */
```
Comment (bgrins) [inline devtools/client/webconsole/new-console-output/test/fixtures/stubs/pageError.js:2]: The boilerplate in these files should include a comment with something like:

// THIS FILE IS AUTOGENERATED.  DO NOT MODIFY BY HAND.  RUN TESTS IN FIXTURES/ TO UPDATE.

Comment (linclark) [inline devtools/client/webconsole/new-console-output/test/fixtures/stubs/pageError.js:2]: Good call, updated

File: devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/browser_webconsole_update_stubs_evaluation_result.js [hunk 2]
```diff
@@ -0,0 +1,243 @@
+/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
+/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
+/* Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/ */
+
+"use strict";
+
+var { DebuggerServer } = require("devtools/server/main");
```
Comment (bgrins) [inline devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/browser_webconsole_update_stubs_evaluation_result.js]: So, I didn't know about target.activeConsole until just now, but this test is equivalent:

```
Cu.import("resource://gre/modules/osfile.jsm");
const TEST_URI = "data:text/html;charset=utf-8,stub generation";

const { evaluationResult: snippets} = require("devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/stub-snippets.js");

let stubs = [];

add_task(function* () {
  let toolbox = yield openNewTabAndToolbox(TEST_URI, "webconsole");
  ok(true, "make the test not fail");

  for (var [code,key] of snippets) {
    const packet = yield new Promise(resolve => {
      toolbox.target.activeConsole.evaluateJS(code, resolve);
    });
    stubs.push(formatStub(key, packet));
    if (stubs.length == snippets.size) {
      let filePath = OS.Path.join(`${BASE_PATH}/stubs`, "evaluationResult.js");
      OS.File.writeAtomic(filePath, formatFile(stubs));
    }
  }
});
```

File: devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/test_webconsole_update_stubs_page_error.html [hunk 3]
```diff
@@ -0,0 +1,88 @@
+<!DOCTYPE HTML>
```
Comment (bgrins) [inline devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/test_webconsole_update_stubs_page_error.html]: Found a way to convert this into a mochitest-browser and get rid of the common.js and extra test file boilerplate.  Something like:

```
Cu.import("resource://gre/modules/osfile.jsm");
const TEST_URI = "data:text/html;charset=utf-8,stub generation";

const { pageError: snippets} = require("devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/stub-snippets.js");

let stubs = [];

add_task(function* () {
  let toolbox = yield openNewTabAndToolbox(TEST_URI, "webconsole");
  ok(true, "make the test not fail");

  for (var [key,code] of snippets) {
    let received = new Promise(resolve => {
      toolbox.target.client.addListener("pageError", function onPacket(e, packet) {
        toolbox.target.client.removeListener("pageError", onPacket);
        info("Received page error:" + e + " " + JSON.stringify(packet, null, "\t"));

        let message = prepareMessage(packet, {getNextId: () => 1});
        stubs.push(formatStub(message.messageText, packet));
        resolve();
      });
    });

    info("Injecting script: " + code);

    yield ContentTask.spawn(gBrowser.selectedBrowser, code, function(code) {
      let container = content.document.createElement("script");
      content.document.body.appendChild(container);
      container.textContent = code;
      content.document.body.removeChild(container);
    });

    yield received;
  }

  let filePath = OS.Path.join(`${BASE_PATH}/stubs`, "pageError.js");
  OS.File.writeAtomic(filePath, formatFile(stubs));
});
```

**Comment by linclark** [comment]
@bgrins fixes up. are you ok with me landing this?

File: devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/browser_webconsole_update_stubs_console_api.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    let hud = toolbox.getCurrentPanel().hud;
+    let {ui} = hud;
+
+    ok(ui.jsterm, "jsterm exists");
+    ok(ui.newConsoleOutput, "newConsoleOutput exists");
+
+    toolbox.target.client.addListener("consoleAPICall", (type, res) => {
+      stubs.push(formatStub(key, res));
+      if (stubs.length == snippets.size) {
+        let filePath = OS.Path.join("../../../../devtools/client/webconsole/new-console-output/test/fixtures/stubs", "consoleApi.js");
```
Comment (bgrins) [inline devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/browser_webconsole_update_stubs_console_api.js]: Should this be `${BASE_PATH}/stubs`?

**Comment by bgrins** [comment]
> @bgrins fixes up. are you ok with me landing this?

Yes, feel free to land

**Pull request merged**